File: cpp-client/src/Chain.cpp
#include "Chain.hpp"

#include <cstdlib>
#include <stdexcept>
#include <nlohmann/json.hpp>

#include "Crypto.hpp"   // crypto::to_base64
#include "Keccak.hpp"

using json = nlohmann::json;

// First 4 bytes of keccak256("getKey(bytes32)") — fixed by the contract
// ABI. Hardcoded constant; the Keccak implementation itself is
// vector-tested separately.
static const char* GET_KEY_SELECTOR = "12aaac70";

ChainConfig ChainConfig::from_env() {
    ChainConfig cfg;
    if (const char* url = std::getenv("SECUREMAILBOX_RPC_URL")) cfg.rpc_url = url;
    if (const char* addr = std::getenv("SECUREMAILBOX_KEY_REGISTRY")) cfg.key_registry_address = addr;
    return cfg;
}

ChainClient::ChainClient(ChainConfig config)
    : config_{std::move(config)}
    , curl_{curl_easy_init()}
{
    if (!curl_) {
        throw std::runtime_error("curl_easy_init() failed — is libcurl installed?");
    }
}

size_t ChainClient::write_callback(char* ptr, size_t size, size_t nmemb, void* userdata) {
    auto* body = static_cast<std::string*>(userdata);
    body->append(ptr, size * nmemb);
    return size * nmemb;
}

std::string ChainClient::rpc_post(const std::string& body, std::string& error) {
    std::string response;

    curl_easy_reset(curl_.get());
    curl_easy_setopt(curl_.get(), CURLOPT_URL, config_.rpc_url.c_str());
    curl_easy_setopt(curl_.get(), CURLOPT_WRITEFUNCTION, write_callback);
    curl_easy_setopt(curl_.get(), CURLOPT_WRITEDATA, &response);
    curl_easy_setopt(curl_.get(), CURLOPT_POST, 1L);
    curl_easy_setopt(curl_.get(), CURLOPT_POSTFIELDS, body.c_str());
    curl_easy_setopt(curl_.get(), CURLOPT_POSTFIELDSIZE, static_cast<long>(body.size()));
    curl_easy_setopt(curl_.get(), CURLOPT_TIMEOUT, 15L);

    curl_slist* headers = curl_slist_append(nullptr, "Content-Type: application/json");
    curl_easy_setopt(curl_.get(), CURLOPT_HTTPHEADER, headers);

    const CURLcode res = curl_easy_perform(curl_.get());
    curl_slist_free_all(headers);

    if (res != CURLE_OK) {
        error = std::string("RPC unreachable: ") + curl_easy_strerror(res);
        return {};
    }
    long status = 0;
    curl_easy_getinfo(curl_.get(), CURLINFO_RESPONSE_CODE, &status);
    if (status != 200) {
        error = "RPC returned HTTP " + std::to_string(status);
        return {};
    }
    return response;
}

OnChainKey ChainClient::get_onchain_key(const std::string& username) {
    OnChainKey out;

    if (config_.key_registry_address.empty()) {
        out.reason = "KeyRegistry address not configured on this client.";
        return out;
    }

    // calldata = selector ‖ keccak256(username), all computed locally.
    const std::string data = std::string("0x") + GET_KEY_SELECTOR + keccak::keccak256_hex(username);

    const json req = {
        {"jsonrpc", "2.0"},
        {"id", 1},
        {"method", "eth_call"},
        {"params", {{{"to", config_.key_registry_address}, {"data", data}}, "latest"}},
    };

    std::string error;
    const std::string body = rpc_post(req.dump(), error);
    if (!error.empty()) {
        out.reason = error;
        return out;
    }

    try {
        const json resp = json::parse(body);
        if (resp.contains("error")) {
            out.reason = "RPC error: " + resp["error"].dump();
            return out;
        }

        std::string hex = resp.at("result").get<std::string>();
        if (hex.rfind("0x", 0) == 0) hex = hex.substr(2);

        // getKey returns (bytes32, uint32, uint64, bool): four static
        // 32-byte words, decoded by slicing — no ABI library needed.
        if (hex.size() != 4 * 64) {
            out.reason = "Unexpected eth_call result length " + std::to_string(hex.size());
            return out;
        }

        const std::string key_hex = hex.substr(0, 64);
        const unsigned version =
            static_cast<unsigned>(std::stoul(hex.substr(64, 64), nullptr, 16));
        const std::uint64_t updated_at = std::stoull(hex.substr(128, 64), nullptr, 16);
        const bool revoked = std::stoul(hex.substr(192, 64), nullptr, 16) == 1;

        out.ok = true;
        if (version == 0) {
            out.registered = false;
            return out;
        }

        std::vector<unsigned char> key_bytes(32);
        for (std::size_t i = 0; i < 32; ++i) {
            key_bytes[i] = static_cast<unsigned char>(
                std::stoul(key_hex.substr(i * 2, 2), nullptr, 16));
        }

        out.registered = true;
        out.version    = version;
        out.key_b64    = crypto::to_base64(key_bytes);
        out.updated_at = updated_at;
        out.revoked    = revoked;
        return out;
    } catch (const std::exception& e) {
        out.ok = false;
        out.reason = std::string("Failed to parse RPC response: ") + e.what();
        return out;
    }
}

File: web-client/js/chain.js
/**
 * chain.js — direct client → Sepolia reads of the KeyRegistry contract.
 *
 * SECURITY MODEL (docs/crypto-design.md §3(d)1, §8.11): this module is the
 * client side of the on-chain key-transparency check. Everything is
 * computed locally — the identity hash (keccak256 of the username), the
 * calldata, and the decode of the result — and the JSON-RPC request goes
 * straight to a public Sepolia endpoint, never through the mailbox server.
 * Routing this lookup through the server would let a compromised server
 * answer its own integrity check, which is precisely the attack the
 * registry exists to expose.
 *
 * Failure semantics are the CALLER's job and differ by call site:
 * the pre-encrypt gate must FAIL CLOSED on any RPC failure (an active
 * network attacker who can block the RPC must not be able to silently
 * disable the check), so this module never swallows errors — it returns
 * either a decoded record or a status:'error' result the caller must act on.
 */

import { CHAIN } from './config.js';
import { keccak256 } from './keccak.js';

// First 4 bytes of keccak256("getKey(bytes32)") — fixed by the contract ABI.
// (Constant, so it is hardcoded rather than recomputed; the Keccak
// implementation itself is vector-tested separately.)
const GET_KEY_SELECTOR = '12aaac70';

const bytesToHex = (bytes) =>
  Array.from(bytes).map((b) => b.toString(16).padStart(2, '0')).join('');

/** keccak256(username) — the registry's identity scheme, computed locally. */
export function identityHash(username) {
  return bytesToHex(keccak256(new TextEncoder().encode(username)));
}

/**
 * Read KeyRegistry.getKey(keccak256(username)) directly over JSON-RPC.
 *
 * @returns one of:
 *   { status: 'ok', registered: boolean, version: number,
 *     keyB64: string|null, updatedAt: number|null, revoked: boolean }
 *   { status: 'error', reason: string }   — RPC unreachable, bad response,
 *     or registry address not configured. The pre-encrypt gate treats this
 *     as a hard stop (fail closed, explicit typed override only).
 */
export async function getOnChainKey(username) {
  if (!CHAIN.keyRegistryAddress) {
    return { status: 'error', reason: 'KeyRegistry address not configured on this client.' };
  }

  const data = '0x' + GET_KEY_SELECTOR + identityHash(username);

  let res;
  try {
    res = await fetch(CHAIN.rpcUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        jsonrpc: '2.0',
        id: 1,
        method: 'eth_call',
        params: [{ to: CHAIN.keyRegistryAddress, data }, 'latest'],
      }),
    });
  } catch (err) {
    return { status: 'error', reason: `RPC unreachable: ${err.message}` };
  }
  if (!res.ok) {
    return { status: 'error', reason: `RPC returned HTTP ${res.status}` };
  }

  let body;
  try {
    body = await res.json();
  } catch {
    return { status: 'error', reason: 'RPC returned a non-JSON response' };
  }
  if (body.error) {
    return { status: 'error', reason: `RPC error: ${body.error.message ?? JSON.stringify(body.error)}` };
  }

  // getKey returns (bytes32 x25519Key, uint32 version, uint64 updatedAt,
  // bool revoked) — four static 32-byte words, decoded by slicing; no ABI
  // library needed for a fixed layout.
  const hex = (body.result ?? '').replace(/^0x/, '');
  if (hex.length !== 4 * 64) {
    return { status: 'error', reason: `Unexpected eth_call result length ${hex.length}` };
  }

  const keyHex = hex.slice(0, 64);
  const version = parseInt(hex.slice(64, 128), 16);
  const updatedAt = parseInt(hex.slice(128, 192), 16);
  const revoked = parseInt(hex.slice(192, 256), 16) === 1;

  if (version === 0) {
    return { status: 'ok', registered: false, version: 0, keyB64: null, updatedAt: null, revoked: false };
  }

  const keyBytes = new Uint8Array(keyHex.match(/.{2}/g).map((h) => parseInt(h, 16)));
  return {
    status: 'ok',
    registered: true,
    version,
    keyB64: btoa(String.fromCharCode(...keyBytes)),
    updatedAt,
    revoked,
  };
}
